wheel/python/clvm_rs/program.py

```diff
@@ -0,0 +1,350 @@
+from __future__ import annotations
+from typing import Iterator, List, Tuple, Optional, BinaryIO
+
+from .at import at
+from .casts import CastableType, to_clvm_object, int_from_bytes, int_to_bytes
+from .chia_dialect import CHIA_DIALECT
+from .clvm_rs import run_serialized_chia_program
+from .clvm_storage import CLVMStorage
+from .clvm_tree import CLVMTree
+from .curry_and_treehash import CurryTreehasher
+from .eval_error import EvalError
+from .replace import replace
+from .ser import sexp_from_stream, sexp_to_stream, sexp_to_bytes
+from .tree_hash import sha256_treehash
+
+
+
+class Program(CLVMStorage):
+    """
+    A wrapper around `CLVMStorage` providing many convenience functions.
+    """
+
+    UNSAFE_MAX_COST: Optional[int] = None
+
+    curry_treehasher: CurryTreehasher = CurryTreehasher(CHIA_DIALECT)
+    _cached_serialization: Optional[bytes]
+
+    # serialization/deserialization
+
+    @classmethod
+    def parse(cls, f: BinaryIO) -> Program:
+        return sexp_from_stream(f, cls.new_pair, cls.new_atom)
```

Comment:
`parse()`, `from_bytes()` and `stream()` resemble the protocol we use for `Streamable` types. It doesn't seem appropriate to use this type in a streamable structure though, since this function, `parse()` looks quite slow.
maybe it could be made faster by making a copy of the bytes and then use the clvm tree method of parsing.
I suppose there would have to be some glue in `Streamable`, which could call `from_bytes_with_cursor()` instead.

Reply:
I agree this is still slow. I have a version of this `parse` function in a separate branch that addresses this issue. It wraps the `BinaryIO` to `tee` the blob to a `io.BytesIO` as things are pulled out of it, and then quickly skips over all the bytes using a simple `0xff` counter algorithm like that in `SerializedProgram`. Now we've turned the streamed program into a blob, and we can parse it using `from_bytes`. There are python and rust implementations, although both need to repeatedly call `f.read` from python to get bytes out.

I didn't include it in this patch because I'm not completely happy with it yet, and this patch is already large and complex enough. Baby steps.